sbt-metals plugin

laughedelic opened the pull request:
This will fix #219. It's based on the WIP #229. Once #229 is ready I will rebase this branch. 

- [x] move `MetalsPlugin.scala` and add a subproject for it
- [x] refactor MetalsPlugin code:
    + separate sbt plugin declarations from the tasks/commands implementations
    + clarify tasks scoping: now
        * `lsp4s / Test / metalsWriteBuildinfo` generates only `lsp4s/test.properties` (exact project + configuration scope)
        * `lsp4s / metalsWriteBuildinfo` generates only `lsp4s/{compile,test}.properties` (aggregates over configurations)
        * `Global / metalsWriteBuildinfo` aggregates over all subprojects and their configurations (`metalsSetup` uses it)
    + default configuration scopes are `Compile` and `Test`, but user can add `metalsSettings(Compile, Foo, Bar)` for extra configurations

gabro commented on build.sbt:
```diff
@@ -95,15 +89,15 @@ lazy val noPublish = List(
+  .in(file("."))
+  .settings(
+    noPublish
+  )
```
can you preserve `name := "metalsRoot"` here?

The reason why it was there is that sbt uses a bogus name for the root project, because the default name (the directory) is already taken by a project.

laughedelic replied:
I don't know why exactly, but when I started working on this PR, `name := "metalsRoot"` didn't work and that bodus name appeared, so I added this explicit root project declaration. Now it's called `root`, which I think is quite standard naming in multi-build projects. But if you prefer `metalsRoot`, I can rename it, no problem.

olafurpg replied:
`metalsRoot` is better since this name appears in IntelliJ "Open recent"

![screen shot 2018-04-03 at 15 43 13](https://user-images.githubusercontent.com/1408093/38252970-c72da9fe-3755-11e8-9930-586f3c1191da.png)

gabro replied:
ah, I see.
No problem, as long as the horrible generated name doesn't show up, I'm fine with "root"!

gabro replied:
Nevermind, what @olafurpg said :D

laughedelic replied:
OK 👌

gabro commented on build.sbt:
```diff
@@ -181,3 +175,18 @@ lazy val testWorkspace = project
+      Seq("-Yrangepos") ++
+        Seq("-Ywarn-unused-import").filter { _ =>
+          scalaBinaryVersion.value == "2.10"
+        }
```
I find this whole `diff` pretty hard to read. I took me a while to understand.

Doesn't this work?

```scala
scalacOptions --= Seq("-Yrangepos", "-Ywarn-unused-import")
```

laughedelic replied:
I admit that this `diff` doesn't look nice. I wanted to remove `-Ywarn-unused-import` only on Scala 2.10, but I guess it's not important, I'll change it to `--=` 👍

gabro replied:
you can apply the filter on that `Seq` too, but unless it's necessary I would avoid complicating it.

laughedelic commented:
Ready 👌 I rebased it on top of #229. After that PR is merged I'll update and it will be easier to review. But in general all changes are only in `build.sbt` and `sbt-metals/src/...`. I updated PR description with the summary of the plugin refactoring. From the user perspective nothing has changed (since #229): just run `metalsSetup`!

> @olafurpg Does this destroy automatic aggregation?

Yes, I didn't realize that before, I changed it back. I got confused by the output of `projects`, it still shows an autogenerated id `default-eaa2d5` for the root project. But if `name :=` works for you, I'm fine with it.

gabro commented:
@laughedelic Now that #229 is in, I'll review as soon as it's rebased, for simplicity.

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -1,91 +1,91 @@
 import java.io._
-import sbt.Def
 
 object MetalsPlugin extends AutoPlugin {
```
Can we move this file to `scala/meta/sbt/MetalsPlugin.scala`? Or else IntelliJ complains about a wrong location. I think it's a good practice to match directory with package.

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -98,16 +98,16 @@ object MetalsPlugin extends AutoPlugin {
       configDir.mkdirs()
       "semanticdbEnable" ::
-        "*:metalsWriteCompilerConfig" ::
+        "Global / metalsWriteCompilerConfig" ::
```
Will this work in sbt 0.13? We should try to keep 0.13 compatibility in MetalsPlugin.scala

gabro replied:
Doesn't the CI check this? `^compile` should catch it, right?

laughedelic replied:
Right, it should be `*/metalsWriteCompilerConfig`. 

@gabro it's just a string, so it compiles 👌 but if you run it on sbt 0.13, it will fail 🌵 We should add scripted tests for the plugin to avoid such mistakes (but in another PR already)

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -1,91 +1,91 @@
   override def requires = sbt.plugins.JvmPlugin
   object autoImport {
 
     val metalsCompilerConfig =
```
Yak shaving. How about calling this `metalsBuildInfo`? To match `.metals/buildinfo`

laughedelic replied:
That's strange, I actually did it in 2c61ebc, except the case: `metalsBuildinfo`. I will rename it to `..Info`. I would also suggest to raneme all related things in the server code.

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -1,91 +1,91 @@
     val metalsCompilerConfig =
-      taskKey[String](
+      taskKey[Seq[(String, String)]](
         "String containing build metadata in properties file format."
```
> List of key/value pairs for build information such as classpath/sourceDirectories.

laughedelic replied:
Again, this refers to an outdated diff (sorry for the mess)

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -1,91 +1,91 @@
 
     val metalsCompilerConfig =
-      taskKey[String](
+      taskKey[Seq[(String, String)]](
```
How about `Map[String, String]` to better convey the structure?

laughedelic replied:
I did that first, but sbt shows `Seq` nicer if you call `show metalsCompilerConfig`. On the other hand it's still not very readable, since values are huge. So I'll change it to Map again.

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -1,91 +1,91 @@
+            ),
+            metalsWriteCompilerConfig := {
+              val props = new java.util.Properties()
+              metalsCompilerConfig.value.foreach {
```
Nice 👍

olafurpg commented on sbt-metals/src/main/scala/MetalsPlugin.scala:
```diff
@@ -98,16 +98,16 @@ object MetalsPlugin extends AutoPlugin {
-    }.value
+    // without project scope it will aggregate over all projects
+    metalsWriteCompilerConfig :=
+      metalsWriteCompilerConfig.all(ScopeFilter(inAnyProject)).value
```
Will this run for `Test` config?

olafurpg commented:
> lsp4s / metalsWriteBuildinfo generates only lsp4s/{compile,test}.properties (aggregates over configurations)

Wow, I didn't know it worked like that. sbt configurations are a mystery to me.

I will need to give the new plugin a try locally

gabro commented:
> Hmm, maybe "single MetalsPlugin.scala file" is no longer worth it if we pubish the plugin 🤔

It is in case you're behind a corporate proxy and you're forced to copy-paste the file, at least according to Sam.

olafurpg commented:
> It is in case you're behind a corporate proxy and you're forced to copy-paste the file, at least according to Sam.

In that case I propose to concat the two files. I think it's worth keeping "single file install" for at least a while longer until the sbt plugin has been released on maven.

laughedelic commented:
Addressed feedback and rebased on master. I'm still not sure how to go about publishing. What we need (as I see it) is to run

1. run `+releaseEarly` but excluding `sbt-metals` subproject
2. run `; project sbt-metals; ^releaseEarly` to cross-publish the plugin

One way to do the 1st step is to create a root project which explicitly aggregates all subprojects except sbt-metals and run `releaseEarly` on it. Then add a `release` command which runs the 2 steps.

Does anyone have a better idea?

gabro commented:
> One way to do the 1st step is to create a root project which explicitly aggregates all subprojects except sbt-metals and run releaseEarly on it.

I think that's the best way to go. It gives us the flexibility that we will eventually need.

> Then add a release command which runs the 2 steps

This is not strictly necessary: the CI can just run the two steps directly.

gabro commented on sbt-metals/src/main/scala/scala/meta/sbt/MetalsPlugin.scala:
```diff
@@ -0,0 +1,189 @@
+    // without project scope it will aggregate over all projects
+    metalsWriteBuildInfo := {
+      metalsWriteBuildInfo.all(ScopeFilter(inAnyProject)).value
+      streams.value.log.info("Metals rocks! 🤘")
```
😆

laughedelic replied:
We already have some emoji in the log from `semanticdbEnable`, so I thought, why not? I hope it will make some users smile when they first try Metals... 😊

olafurpg replied:
I like emojis 👍 

How about going all in and making it more consistent? 😄 


> 👌 semanticdb-scalac enabled
> 🔨 created .metals/buildinfo/foo/test.properties
> 🤘 Metals is ready, time to rock!

laughedelic commented:
@jvican we need some sbt-guru help here. Something is wrong with the `^` in the plugin subproject. It changes `sbtVersion in pluginCrossBuild`, but then compiles only for sbt 1.0:

```
[sbt-1.1.2] org.scalameta/metalsRoot:sbt-metals v0.1-SNAPSHOT » project sbt-metals
[info] Set current project to sbt-metals (in build file:/Users/laughedelic/dev/laughedelic/metals/)
[sbt-1.0.4] org.scalameta/sbt-metals:sbt-metals v0.1-SNAPSHOT » ^compile
[info] Setting `sbtVersion in pluginCrossBuild` to 0.13.17
[info] Set current project to sbt-metals (in build file:/Users/laughedelic/dev/laughedelic/metals/)
[info] Compiling 1 Scala source to /Users/laughedelic/dev/laughedelic/metals/sbt-metals/target/scala-2.12/sbt-1.0/classes ...
[info] Done compiling.
[success] Total time: 9 s, completed Apr 4, 2018 6:52:44 PM
[info] Setting `sbtVersion in pluginCrossBuild` to 1.0.4
[info] Set current project to sbt-metals (in build file:/Users/laughedelic/dev/laughedelic/metals/)
[success] Total time: 0 s, completed Apr 4, 2018 6:52:46 PM
[info] Setting `sbtVersion in pluginCrossBuild` to 1.1.2
[info] Set current project to sbt-metals (in build file:/Users/laughedelic/dev/laughedelic/metals/)
[sbt-1.0.4] org.scalameta/sbt-metals:sbt-metals v0.1-SNAPSHOT » show crossSbtVersions
[info] * 0.13.17
[info] * 1.0.4
```

And setting version manually with `^^` doesn't help either:

```
[sbt-1.0.4] org.scalameta/sbt-metals:sbt-metals v0.1-SNAPSHOT » ^^0.13.17
[info] Setting `sbtVersion in pluginCrossBuild` to 0.13.17
[info] Set current project to sbt-metals (in build file:/Users/laughedelic/dev/laughedelic/metals/)
[sbt-1.0.4] org.scalameta/sbt-metals:sbt-metals v0.1-SNAPSHOT » clean
[success] Total time: 0 s, completed Apr 4, 2018 6:53:32 PM
[sbt-1.0.4] org.scalameta/sbt-metals:sbt-metals v0.1-SNAPSHOT » compile
[info] Compiling 1 Scala source to /Users/laughedelic/dev/laughedelic/metals/sbt-metals/target/scala-2.12/sbt-1.0/classes ...
[info] Done compiling.
[success] Total time: 3 s, completed Apr 4, 2018 6:53:37 PM
```

Could you take a look at it?

jvican commented:
I've run into the same problem before, it happens because `^` takes the sbt version from the `crossSbtVersion`s of the root project, not the sbt plugin project. https://github.com/scalacenter/bloop/blob/master/build.sbt#L113-L120

olafurpg commented on bin/release.sh:
```diff
@@ -0,0 +1,9 @@
+
+# release only on merge commits or tags
+if [[ "$(git rev-list --merges HEAD^..HEAD)" || "$TRAVIS_TAG" ]]; then
+    sbt '+releaseEarly' 'project sbt-metals' '^releaseEarly'
```
This can be automated with a `commands += Command.command("release")(s => "a" :: "b" :: s)`

laughedelic replied:
I just followed @gabro's advice here https://github.com/scalameta/metals/pull/230#issuecomment-378645212. But I'll add a command, just to keep sbt-specific things in within sbt build.

laughedelic commented:
@jvican thanks! adding `crossSbtVersions` to the root project and setting explicitly `scalaVersion` in the plugin project solved the problem. But this is definitely not cool 👎 is this problem tracked in sbt?

@olafurpg I added the command and improved log messages, but didn't add 🔨 because I think it's too much and hammer's handle doesn't look good on the dark background:

<img width="601" alt="screen shot 2018-04-05 at 01 21 22" src="https://user-images.githubusercontent.com/766656/38339669-c0ae091e-386f-11e8-9807-d5b873154cca.png">

I tried a couple of other emojis, but I think it's just too much. Do you have other suggestions?

Also let me know if aggregation now works for you as intended.

One inconvenience with having source dependency is that it gets overridden by the version added globally.

We definitely need scripted tests for this, but in another PR (I'm a bit tired of sbt quirks now 😒).

olafurpg commented on sbt-metals/src/main/scala/scala/meta/sbt/MetalsPlugin.scala:
```diff
@@ -0,0 +1,189 @@
+    buildinfoDir.mkdirs()
+    val outFile = buildinfoDir / s"${configuration.value.name}.properties"
+    IO.write(outFile, out.toString())
+    streams.value.log.info(s"Created ${outFile.relativeTo(baseDir).get}")
```
Can we print out the absolute path to `outFile`? I just tested this locally and realized with relative paths you can't Cmd+click on the path in iTerm to open the file. I think we should make it easy to poke into the files for curious users.

gabro replied:
minor non-blocking comment on this too: Is `Created` always accurate? Don't these get `Updated` sometimes? If that's the case we could get away with `Written` :P

olafurpg replied:
> Wrote /path/to/foo

laughedelic replied:
😄👍 I agree. I noticed that sbt itself is very inconsistent in this regard. It says `Writing Ivy file ...xml`, but then `Wrote ...pom` 😬 . 

So "Written" or "Wrote" or "Writing"?

gabro replied:
Slightly longer suggestion

| Written build information for lsp4s/compile to /path/to/foo

laughedelic replied:
I think this is will be too long with absolute paths and quite repetitive:

```
Written build information for integration/compile to /Users/laughedelic/dev/laughedelic/metals/.metals/buildinfo/integration/compile.properties
```

gabro replied:
It's a bit long I agree. My intention was to vertically align the sbt projects/configurations so to easily spot what was included.

laughedelic replied:
The path prefix is the same, so they will be aligned. But you should also take into account that these logs are not ordered (sbt tasks aggregation runs in parallel).

gabro commented:
🚀
